File: mysql-async/src/main/java/com/github/mauricio/async/db/mysql/encoder/AuthenticationSwitchResponseEncoder.scala
package com.github.mauricio.async.db.mysql.encoder

import com.github.mauricio.async.db.mysql.message.client.{AuthenticationSwitchResponse, ClientMessage}
import io.netty.buffer.ByteBuf
import com.github.mauricio.async.db.exceptions.UnsupportedAuthenticationMethodException
import com.github.mauricio.async.db.mysql.encoder.auth.AuthenticationMethod
import java.nio.charset.Charset
import com.github.mauricio.async.db.util.ByteBufferUtils

class AuthenticationSwitchResponseEncoder( charset : Charset ) extends MessageEncoder {

  def encode(message: ClientMessage): ByteBuf = {
    val switch = message.asInstanceOf[AuthenticationSwitchResponse]

    val method = switch.request.method
    val authenticator = AuthenticationMethod.Availables.getOrElse(
    method, { throw new UnsupportedAuthenticationMethodException(method) })

    val buffer = ByteBufferUtils.packetBuffer()

    val bytes = authenticator.generateAuthentication(charset, switch.password, switch.request.seed.getBytes(charset) )
    buffer.writeBytes(bytes)

    buffer
  }

}

File: mysql-async/src/main/java/com/github/jasync/sql/db/mysql/encoder/AuthenticationSwitchResponseEncoder.kt
package com.github.jasync.sql.db.mysql.encoder

import com.github.jasync.sql.db.exceptions.UnsupportedAuthenticationMethodException
import com.github.jasync.sql.db.mysql.encoder.auth.AuthenticationMethod
import com.github.jasync.sql.db.mysql.message.client.AuthenticationSwitchResponse
import com.github.jasync.sql.db.mysql.message.client.ClientMessage
import com.github.jasync.sql.db.util.ByteBufferUtils
import io.netty.buffer.ByteBuf
import java.nio.charset.Charset

class AuthenticationSwitchResponseEncoder(val charset: Charset) : MessageEncoder {

    override fun encode(message: ClientMessage): ByteBuf {
        val switch = message as AuthenticationSwitchResponse

        val method = switch.request.method
        val authenticator = AuthenticationMethod.Availables.getOrElse(method) {
            throw UnsupportedAuthenticationMethodException(method)
        }

        val buffer = ByteBufferUtils.packetBuffer()

        val bytes = authenticator.generateAuthentication(
            charset,
            switch.password,
            switch.request.seed,
            switch.sslConfiguration,
            switch.rsaPublicKey,
        )
        buffer.writeBytes(bytes)

        return buffer
    }
}
